README
Overview
--------

This package provides an API for accessing departure data from the realtime
system of the Oslo public transportation system. Its features are:

- Find stations by name
- Retrieve realtime data for stations. Falls back to schedule data when no
  realtime data is available

In addition to the API, two command line scripts are included: tf_realtime
and tf_search.

Contacts
--------
Author: runefh at gmail dot com
Web site: http://bitbucket.org/runeh/pytrafikanten/

License and Legal stuff
-----------------------

The library is distributed under the BSD license. See the LICENSE file
The realtime data is distributed under the Attribution-Noncommercial-Share
Alike 3.0 Unported ( http://creativecommons.org/licenses/by-nc-sa/3.0/deed.en )

More information at http://labs.trafikanten.no/lisens-for-bruk-av-trafikanten-data-til-ikke-kommersiell-bruk.php

For inquries about commercial use, contact trafikanten. See:
http://labs.trafikanten.no/om-labs/kontakt.php

Also note that in the case of abuse, or excessive popularity, trafikanten may
choose to disable the web service if it is disrupting their normal services.

Example
-------

Import the module:

    >>> import trafikanten

Search for some stations:

    >>> trafikanten.find_station("holmenk")
    [{'xcoord': u'592869', 'ycoord': u'6648156', 'id': u'03012370', 'district': u'Oslo', 'name': u'Holmenkollen'}]

Get realtime data for the found station:

    >>> trafikanten.get_realtime('03012370') # doctest: +ELLIPSIS
    [...]

The same usign a caching class would go like this:

    >>> import trafikanten
    >>> tf = trafikanten.MemoryCacheTrafikanten()
    >>> tf.find_station("holmenk")
    [{'xcoord': u'592869', 'ycoord': u'6648156', 'id': u'03012370', 'district': u'Oslo', 'name': u'Holmenkollen'}]
    >>> tf.get_realtime('03012370') # doctest: +ELLIPSIS
    [...]

Usage notes
------------

The package provdes module level functions for making requests to trafikanten
without any caching. It's however strongly recommended that developers use
one of the provided caching API classes, or implement their own caching.

With regard to caching, the following rules of thumb should be followed:

- Station search data can be cached for a long time, as that data rarely
    changes. An hour is the default in the API, but more is fine.
- Realtime data should be cached for no more than 20 seconds. In other words,
    this caching will only be useful under load.

The above rules of thumb are the defaults for the caching classes.

Known issues
------------

- There is currently no way to map from a station id to a station name. If you
  need this your only choice is to cache all station id/name pairs from
  searches and hope whatever you're looking for has been searched for at
  one point.
- There is currently no way to determine if realtime data is available for a
  certain stop. Getting an empty list as the result from a realtime query may
  mean that there is no data available, or that there are noe upcomming
  departures.
- There is currently no way to get the "shortname" of a station, or to retrieve
  realtime data based on shortnames

To do
-----

- More tests
- More docs
- Better caching classes


CHANGELOG
0.1

    * Initial release

0.2

    * Added custom user-agent string
    * Added lat/lon to station data (Patch from Marius Mathiesen)


tests/tests.py
# coding=utf-8

# fixme : test for unicode everywhere!


import os
import types
import trafikanten
import trafikanten.util
import time

def smoke_test_realtime_parser():
    folder = "tests/sample_realtime"
    files = os.listdir(folder)

    # make sure that all the realtime data files in the sample folder is
    # parsed without excepting and returns one or more results
    for path in [os.path.join(folder, f) for f in files]:
        s = open(path).read()
        assert trafikanten.api._parse_realtime_data(s) != None, "Oh noes! %s" % path


def smoke_test_search_parser():
    folder = "tests/sample_search"

    files = os.listdir(folder)

    # make sure that all the realtime data files in the sample folder is
    # parsed without excepting and returns one or more results
    for path in [os.path.join(folder, f) for f in files]:
        s = open(path).read()
        assert trafikanten.api._parse_station_data(s) != None, "Oh noes! %s" % path

def test_search_parser_all_strings_unicode():
    folder = "tests/sample_search"
    files = os.listdir(folder)

    # make sure that all the realtime data files in the sample folder is
    # parsed without excepting and returns one or more results
    for path in [os.path.join(folder, f) for f in files]:
        s = open(path).read()
        parsed = trafikanten.api._parse_station_data(s)
        for entry in parsed:
            for key, value in entry.items():
                if isinstance(value, types.StringType):
                    assert isinstance(value, types.UnicodeType), "Wanted unicode, got %s for key %s" % (type(value), key)

def test_realtime_parser_all_strings_unicode():
    folder = "tests/sample_realtime"
    files = os.listdir(folder)

    # make sure that all the realtime data files in the sample folder is
    # parsed without excepting and returns one or more results
    for path in [os.path.join(folder, f) for f in files]:
        s = open(path).read()
        parsed = trafikanten.api._parse_station_data(s)
        for entry in parsed:
            for key, value in entry.items():
                if isinstance(value, types.StringType):
                    assert isinstance(value, types.UnicodeType), "Wanted unicode, got %s for key %s" % (type(value), key)


def test_util_unicode_search():
    res = [
        {'xcoord': '599475', 'ycoord': '6642907', 'id': '03010661',
         'district': 'Oslo', 'name': u'ÅØÆåøæëêã'}
    ]

    data = trafikanten.util.make_human_readable_stations(res)
    assert isinstance(data, types.UnicodeType)

    data = trafikanten.util.make_delimited_stations(res)
    assert isinstance(data, types.UnicodeType)

    data = trafikanten.util.make_delimited_stations(res, header=True)
    assert isinstance(data, types.UnicodeType)

    data = trafikanten.util.make_xml_stations(res)
    assert isinstance(data, types.UnicodeType)

def test_util_unicode_realtime():
    res = [
        {'direction': u'1', 'destination': u'Kjelsås', 'wait_time': 3271,
         'time': (2008, 12, 26, 14, 47, 0, 4, 361, -1), 'id': u'12',
            'is_realtime': True
        }
    ]

    data = trafikanten.util.make_human_readable_realtime(res)
    assert isinstance(data, types.UnicodeType)

    data = trafikanten.util.make_delimited_realtime(res)
    assert isinstance(data, types.UnicodeType)

    data = trafikanten.util.make_delimited_realtime(res, header=True)
    assert isinstance(data, types.UnicodeType)

    data = trafikanten.util.make_xml_realtime(res)
    assert isinstance(data, types.UnicodeType)


def test_fuzzy_departure_unicode():
    t, d = trafikanten.util.fuzzy_departure({"wait_time": 3})
    assert t == trafikanten.util.FUZZY_NOW
    assert d == None

    t, d = trafikanten.util.fuzzy_departure({"wait_time": 120})
    assert t == trafikanten.util.FUZZY_MINS
    assert d == 2

    now = time.localtime()

    t, d = trafikanten.util.fuzzy_departure({"wait_time": 700, "time": now})
    assert t == trafikanten.util.FUZZY_TIME
    assert d != None

def test_invalid_station_id():
    assert trafikanten.get_realtime("does_not_exist") == None

def test_unicode_term_handling():
    term = u"østerås"
    assert trafikanten.find_station(term)


trafikanten/api.py
# coding=utf-8

import urllib
import time
import math
import xml.dom.minidom as minidom
from trafikanten import __version__ as version
from util import utm_to_lat_lng

# url for station search api
_station_url = "http://www5.trafikanten.no/txml/?type=1&stopname=%s"

# url for non-subway realtime data
_non_subway_rt_url = \
    "http://www.sis.trafikanten.no/xmlrtpi/dis/request?DISID=SN$%s"

# url for subway realtime data
_subway_rt_url = \
    "http://www.sis.trafikanten.no:8088/xmlrtpi/dis/request?DISID=SN$%s"

# known subways station ids. Required because subway stations use a different
# realtime data source than other stations.
_subway_stations = [
    "03011930", "03012220", "03011030", "03012365", "02190070", "02190080",
    "02190090", "03010011", "03010020", "03010031", "03010200", "03010360",
    "03010370", "03010600", "03010610", "03010770", "03010780", "03010950",
    "03011010", "03011020", "03011110", "03011120", "03011130", "03011140",
    "03011200", "03011210", "03011220", "03011310", "03011320", "03011330",
    "03011400", "03011410", "03011430", "03011440", "03011450", "03011510",
    "03011520", "03011530", "03011540", "03011610", "03011620", "03011630",
    "03011710", "03011720", "03011730", "03011810", "03011910", "03011920",
    "03011940", "03012000", "03012010", "03012020", "03012030", "03012040",
    "03012100", "03012120", "03012130", "03012210", "03012230", "03012240",
    "03012260", "03012270", "03012280", "03012305", "03012310", "03012315",
    "03012320", "03012325", "03012330", "03012340", "03012345", "03012350",
    "03012355", "03012360", "03012370", "03012375", "03012380", "03012385",
    "03012390", "03012410", "03012420", "03012430", "03012450", "03012460",
    "03012560", "03012565", "03012572", "03012630"
]

class TrafikantentURLopener(urllib.FancyURLopener):
    version = "pytrafikanten/%s" % version

urllib._urlopener = TrafikantentURLopener()

def find_station(term):
    """Search for stations matching term.
    Returns a list of dicts of the form:
    match = {
        "id": "03232323",
        "name": "some station name",
        "district": "district for station",
        "xcoord": "0",
        "ycoord"; "0"
    }
    If no matches where found, the list has length 0. If the search did not
    complete successfully, return None.
    """
    if not term:
        return None

    url = _station_url % urllib.quote(term.encode("utf-8"))
    data = urllib.urlopen(url).read()
    return _parse_station_data(data)

def _get_text(nodelist):
    """Stupid helper for dom. give me .textContent plx
    Returns the contcatentated contents of all text nodes in nodelist"""
    text = u""
    for node in nodelist:
        if node.nodeType == node.TEXT_NODE:
            text = text + unicode(node.data)
    return text

def _single_element(node, name):
    """If node contains one or more children with nodename "name", return the
    first one. If not, return None"""
    elems = node.getElementsByTagName(name)
    if elems: return elems[0]
    else: return None

def _parse_station_data(xmlstr):
    """
    Takes xml a string and returns a list of dicts containing station data.
    """
    doc = minidom.parseString(xmlstr)
    ret = []
    elem_map = {"fromid": "id", "StopName": "name", "District": "district",
                "XCoordinate": "xcoord", "YCoordinate": "ycoord"}
    for elem in doc.getElementsByTagName("StopMatch"):
        entry = {"district": u""}
        for name, value in [ (e.nodeName, _get_text(e.childNodes)) \
                              for e in elem.childNodes \
                              if e.nodeType == e.ELEMENT_NODE  ]:
            if name in elem_map:
                entry[elem_map[name]] = unicode(value)

        if "id" in entry and "name" in entry:
            add_lat_lng_to_entry(entry)
            ret.append(entry)

    return ret

def add_lat_lng_to_entry(entry):
    """Adds a lat and lng for x and y (which is another coordinate system)"""
    entry["lat"], entry["lng"] = utm_to_lat_lng(int(entry['xcoord']), int(entry['ycoord']))

def get_realtime(sid):
    """Get realtime data for station with id "sid". Returns a list of
    dictionaries. If no realtime data is found, the returned list will be
    empty. If the station doesn't exist, or any other error occurs, returns
    None.
    """
    if not sid:
        return None

    if sid in _subway_stations:
        url = _subway_rt_url % sid
    else:
        url = _non_subway_rt_url % sid

    data = urllib.urlopen(url).read()
    result = _parse_realtime_data(data)

    if result:
        result.sort(key=lambda e: int(e["wait_time"]))

    return result

def _parse_realtime_data(xmlstr):
    """
    Takes xml a string and returns a list of dicts containing realtime data.
    """
    doc = minidom.parseString(xmlstr)
    ret = []
    elem_map = {"LineID": "id", "DirectionID": "direction",
                "DestinationStop": "destination" }

    ack = _single_element(doc, "Acknowledge")
    if ack == None or ack.attributes["Result"].nodeValue != "ok":
        return None

    curtime = time.mktime(time.strptime(
        ack.attributes["TimeStamp"].nodeValue[:-10], "%Y-%m-%dT%H:%M:%S"))

    for elem in doc.getElementsByTagName("DISDeviation"):
        entry = {"is_realtime": False}

        for name, value in [ (e.nodeName, _get_text(e.childNodes)) \
                              for e in elem.childNodes \
                              if e.nodeType == e.ELEMENT_NODE ]:
            if name in elem_map:
                entry[elem_map[name]] = unicode(value)
            elif name == "TripStatus":
                entry["is_realtime"] = value == "Real"

        if entry["is_realtime"]:
            timeele = _single_element(elem, "ExpectedDISDepartureTime")
        else:
            timeele = _single_element(elem, "ScheduledDISDepartureTime")

        parsed_time = time.strptime(
            _get_text(timeele.childNodes)[:-10], "%Y-%m-%dT%H:%M:%S")
        entry["time"] = parsed_time
        entry["wait_time"] = int(time.mktime(parsed_time) - curtime)
        ret.append(entry)

    return ret


trafikanten/classes.py
# coding=utf-8

import tempfile
import shutil
import os
import stat
import time
import pickle
import trafikanten
import inspect

class CachingTrafikanten(object):
    """Superclass for classes that want to wrap around trafikanten while
    providing caching. This class should not be instantiated directly, it
    will not work.
    Subclasses need to implement four methods:
    - _get_cached_search(search_term): returns the cached data, or None if not
        in cache (or cache has expired)
    - _cache_search(search_term, data): saves the data for search_term to the
        cache.
    - _get_cached_realtime: same as _get_cached_search for realtime data
    - _cache_realtime: same as _cache_search for realtime data
    The data returned from the _get_* methods should be the same as what is
    returned from trafikanten.find_station and trafikanten.get_realtime.
    """
    def __init__(self, search_expiry_time=3600, realtime_expiry_time=20):
        self.search_expiry_time = search_expiry_time
        self.realtime_expiry_time = realtime_expiry_time

    def find_station(self, term):
        """Identical arguments and return value as trafikanten.find_station.
        Will call _get_cached_search and _cache_search to handle caching
        properly. Subclasses probably do not need to override this method."""
        data = self._get_cached_search(term)
        if data == None:
            data = trafikanten.find_station(term)
            self._cache_search(term, data)

        return data

    def get_realtime(self, sid):
        """Identical arguments and return value as trafikanten.get_realtime.
        Will call _get_cached_realtime and _cache_realtime to handle caching
        properly. Subclasses probably do not need to override this method."""
        data = self._get_cached_realtime(sid)
        if data == None:
            data = trafikanten.get_realtime(sid)
            self._cache_realtime(sid, data)

        return data

    def _get_cached_realtime(self, sid):
        """Return cached realtime data for the station sid. If no cached
        data is available, return None"""
        raise NotImplementedError("%s must be overridden in a subclass." %
                                  inspect.currentframe().f_code.co_name)

    def _cache_realtime(self, sid, data):
        """Cache realtime data for station sid. Returns nothing. This call
        is assumed to always succeed. If implementing a caching backend
        that may fail, this method should probably throw an exception"""
        raise NotImplementedError("%s must be overridden in a subclass." %
                                  inspect.currentframe().f_code.co_name)

    def _get_cached_search(self, term):
        """Get a cached serch data for term. Of the term is not cached, 
        return None"""
        raise NotImplementedError("%s must be overridden in a subclass." %
                                  inspect.currentframe().f_code.co_name)

    def _cache_search(self, term, data):
        """Cache data for search term. Returns nothing. This call
        is assumed to always succeed. If implementing a caching backend
        that may fail, this method should probably throw an exception"""
        raise NotImplementedError("%s must be overridden in a subclass." %
                                  inspect.currentframe().f_code.co_name)

class FileCacheTrafikanten(CachingTrafikanten):
    """Caching version of the trafikanten API that uses the file system for
    caching. It's fairly naive and will not clean up after itself, unless it's
    using a temporary folder and the object gets GCed."""
    def __init__(self, cache_loc=None, **kwargs):
        if cache_loc:
            self.cache_location = cache_loc
            self.uses_temp = False
        else:
            self.cache_location = tempfile.mkdtemp(".temp", "tf_cache.")
            self.uses_temp = True

        CachingTrafikanten.__init__(self, **kwargs)

    def __del__(self):
        """The desctructor just makes sure that if we're using a temporary
        cache directory, it gets deleted on object gc"""
        if self.uses_temp:
            shutil.rmtree(self.cache_location)

    def _get_cached_search(self, term):
        cachepath = os.path.join(self.cache_location, "search_"+str(hash(term)))
        if os.path.isfile(cachepath):
            age = time.time() - os.stat(cachepath)[stat.ST_MTIME]
            if age < self.search_expiry_time:
                fp = open(cachepath, "rb")
                data = pickle.load(fp)
                fp.close()
                return data

        return None

    def _cache_search(self, term, data):
        cachepath = os.path.join(self.cache_location, "search_"+str(hash(term)))
        fp = open(cachepath, "wb")
        pickle.dump(data, fp)
        fp.close()

    def _get_cached_realtime(self, sid):
        cachepath = os.path.join(self.cache_location, "station_"+str(hash(sid)))
        if os.path.isfile(cachepath):
            age = time.time() - os.stat(cachepath)[stat.ST_MTIME]
            if age < self.search_expiry_time:
                fp = open(cachepath, "rb")
                data = pickle.load(fp)
                fp.close()
                return data

        return None

    def _cache_realtime(self, sid, data):
        cachepath = os.path.join(self.cache_location, "search_"+str(hash(sid)))
        fp = open(cachepath, "wb")
        pickle.dump(data, fp)
        fp.close()


class MemoryCacheTrafikanten(CachingTrafikanten):
    """Caching version of the trafikanten API that ram for caching. It's
    fairly naive. Expired data is never removed"""
    def __init__(self, **kwargs):
        self._search_cache = {}
        self._realtime_cache =  {}

        CachingTrafikanten.__init__(self, **kwargs)

    def _get_cached_search(self, term):
        if term in self._search_cache:
            age = time.time() - self._search_cache[term][0]
            if age < self.search_expiry_time:
                return self._search_cache[term][1]

        return None

    def _cache_search(self, term, data):
        self._search_cache[term] = (time.time(), data)

    def _get_cached_realtime(self, sid):
        if sid in self._realtime_cache:
            age = time.time() - self._search_cache[sid][0]
            if age < self.search_expiry_time:
                return self._search_cache[sid][1]

        return None

    def _cache_realtime(self, sid, data):
        self._realtime_cache[sid] = (time.time(), data)


trafikanten/util.py
# coding=utf-8

import math
import time
import xml.dom.minidom as minidom

FUZZY_NOW = 1
FUZZY_MINS = 2
FUZZY_TIME = 3

def fuzzy_departure(dep, mins_cutoff=10, time_format="%H.%M"):
    """Makes a fuzzy representation of the time to wait. Returns a tuple
    containing (result_type, result).

    If the wait time is less than one minute returns (FUZZY_NOW, "Now").
    If the wait time is less than mins_cutoff (in minutes), returns
    (FUZZY_MINS, number_of_minutes_to_wait)
    If the wait time is equal or greater than mins_cutoff returns 
    (FUZZY_TIME, time_for_departure). The time will be formated according
    to time_format, which is in the format of time.striftime."""
    wait = int(dep["wait_time"])
    if wait < 60:
        return (FUZZY_NOW, None)
    elif wait < mins_cutoff*60:
        return (FUZZY_MINS, int(math.floor(wait/60)))
    else:
        return (FUZZY_TIME, time.strftime(time_format, dep["time"]))

def make_human_readable_stations(res):
    """Makes  text representation of a search result that is suitable for 
    showing to a user. Returns a string"""
    ret = []
    if not res:
        ret.append( u"""No stations found""")
    else:
        format = u"%%(id)%ds %%(name)s (%%(district)s)" % \
                                max([ len(e["id"]) for e in res ])
        for dep in res:
            ret.append(format % dep)
    return "\n".join(ret)

def make_delimited_stations(res, header=False, separator=";"):
    """Emit station search result delimited by separator. If header is
    true, include field names as the first line, using same delimiter.
    """
    if not res:
        return ""
    
    separator = unicode(separator)

    format = separator.join(("%(id)s", "%(name)s", "%(district)s",
              "%(xcoord)s", "%(ycoord)s"))
    
    ret = []
    if header:
        ret.append(
            separator.join(("id", "name", "district", "xcoord", "ycoord")))

    for dep in res:
        ret.append(format % dep)
    
    return "\n".join(ret)

def make_xml_stations(res):
    """Serialize as search result in res to xml. Returns xml as a string.
    Format of xml is as follows:
    <stations>
        <station>
            <id>123123</id>
            <name>asdf</name
            <disctrict>Oslo</district>

            fixme: more

    """

    impl = minidom.getDOMImplementation()
    document = impl.createDocument(None, "stations", None)
    body = document.documentElement
    for hit in res:
        station = document.createElement("station")
        for key, val in hit.items():
            ele = document.createElement(key)
            ele.appendChild(document.createTextNode(val))
            station.appendChild(ele)
        body.appendChild(station)
    
    pretty = document.toprettyxml()
    # fixme: minidom docs says this is needed due to bugs with cyclic gc. 
    # Is this still the case though?
    document.unlink() 

    return pretty

def make_delimited_realtime(res, header=False, separator=";"):
    """Emit realtimeresult delimited by separator. If header is true, include
    field names as the first line, using same delimiter
    """
    if not res:
        return ""

    clean_res = [
        { "id": e["id"],
          "destination": e["destination"],
          "direction": e["direction"],
          "is_realtime": str(int(e["is_realtime"])),
          "wait_time": e["wait_time"],
          "time": time.asctime(e["time"])
        } for e in res
    ]

    format = separator.join(("%(id)s", "%(destination)s", "%(direction)s",
              "%(is_realtime)s", "%(wait_time)s", "%(time)s"))
    
    ret = []
    if header:
        ret.append(separator.join(("id", "destination", "direction",
                              "is_realtime", "wait_time", "time")))
    for dep in clean_res:
        ret.append(format % dep)

    return "\n".join(ret)

def make_human_readable_realtime(res):
    if not res:
        return """No departures found for station."""
    else:
        format = u"""%%(id)%ds %%(destination)-%ds %%(formatted_wait)-5s""" % \
                 (max([ len(e["id"]) for e in res ]),
                  max([ len(e["destination"]) for e in res ]))
        ret = []
        for dep in res:
            curdep = {"id": dep["id"], "destination": dep["destination"]}
            fuzz, val = fuzzy_departure(dep)

            if fuzz == FUZZY_NOW:
                curdep["formatted_wait"] = u"Now"
            elif fuzz == FUZZY_MINS and val == 1:
                curdep["formatted_wait"] = u"1 min"
            elif fuzz == FUZZY_MINS:
                curdep["formatted_wait"] = u"%d mins" % val
            else:
                curdep["formatted_wait"] = val
            
            ret.append(format % curdep)

    return "\n".join(ret)

def make_xml_realtime(res):
    """Serialize realtime data in res to xml. Returns xml as a string.
    Format of xml is as follows:
    <realtime>
        <departure>
            <id>123123</id>
            <name>asdf</name
            <time>result from time.asctime</time>
            <is_realtime>yes|no</is_realtime>
            <wait_time>234</wait_time>
            fixme: more

    """
    impl = minidom.getDOMImplementation()

    type_mapper = {
        "time": lambda t: time.asctime(t),
        "wait_time": unicode,
        "is_realtime": lambda b: u"yes" if b else u"no"
    }

    document = impl.createDocument(None, "realtime", None)
    body = document.documentElement
    for hit in res:
        departure = document.createElement("departure")
        for key, val in hit.items():
            ele = document.createElement(key)
            if key in type_mapper:
                ele.appendChild(document.createTextNode(type_mapper[key](val)))
            else:
                ele.appendChild(document.createTextNode(val))
            departure.appendChild(ele)
        body.appendChild(departure)
    
    pretty = document.toprettyxml()

    # fixme: minidom docs says this is needed due to bugs with cyclic gc. 
    # Is this still the case though?
    document.unlink()
    return pretty

def utm_to_lat_lng(easting, northing, zone=32, northernHemisphere=True):
    """Converts from utm to lat_lng. Default zone is the one in which
    Oslo and Akershus falls."""
    if not northernHemisphere:
        northing = 10000000 - northing
    
    a = 6378137
    e = 0.081819191
    e1sq = 0.006739497
    k0 = 0.9996
    
    arc = northing / k0
    mu = arc / (a * (1 - math.pow(e, 2) / 4.0 - 3 * math.pow(e, 4) / 64.0 - 5 * math.pow(e, 6) / 256.0))
    
    ei = (1 - math.pow((1 - e * e), (1 / 2.0))) / (1 + math.pow((1 - e * e), (1 / 2.0)))
    
    ca = 3 * ei / 2 - 27 * math.pow(ei, 3) / 32.0
    
    cb = 21 * math.pow(ei, 2) / 16 - 55 * math.pow(ei, 4) / 32
    cc = 151 * math.pow(ei, 3) / 96
    cd = 1097 * math.pow(ei, 4) / 512
    phi1 = mu + ca * math.sin(2 * mu) + cb * math.sin(4 * mu) + cc * math.sin(6 * mu) + cd * math.sin(8 * mu)
    
    n0 = a / math.pow((1 - math.pow((e * math.sin(phi1)), 2)), (1 / 2.0))
    
    r0 = a * (1 - e * e) / math.pow((1 - math.pow((e * math.sin(phi1)), 2)), (3 / 2.0))
    fact1 = n0 * math.tan(phi1) / r0
    
    _a1 = 500000 - easting
    dd0 = _a1 / (n0 * k0)
    fact2 = dd0 * dd0 / 2
    
    t0 = math.pow(math.tan(phi1), 2)
    Q0 = e1sq * math.pow(math.cos(phi1), 2)
    fact3 = (5 + 3 * t0 + 10 * Q0 - 4 * Q0 * Q0 - 9 * e1sq) * math.pow(dd0, 4) / 24
    
    fact4 = (61 + 90 * t0 + 298 * Q0 + 45 * t0 * t0 - 252 * e1sq - 3 * Q0 * Q0) * math.pow(dd0, 6) / 720
    
    lof1 = _a1 / (n0 * k0)
    lof2 = (1 + 2 * t0 + Q0) * math.pow(dd0, 3) / 6.0
    lof3 = (5 - 2 * Q0 + 28 * t0 - 3 * math.pow(Q0, 2) + 8 * e1sq + 24 * math.pow(t0, 2)) * math.pow(dd0, 5) / 120
    _a2 = (lof1 - lof2 + lof3) / math.cos(phi1)
    _a3 = _a2 * 180 / math.pi
    
    latitude = 180 * (phi1 - fact1 * (fact2 + fact3 + fact4)) / math.pi
    
    if not northernHemisphere:
        latitude = -latitude
    
    longitude = ((zone > 0) and (6 * zone - 183.0) or 3.0) - _a3
    
    return (latitude, longitude)

